File: server/schemas/project.py
# -*- coding: utf-8 -*-
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field


# ---- DefectClass ----
class DefectClassCreate(BaseModel):
    class_index: int = Field(ge=0, description="类别编号 0,1,2…")
    name: str = Field(max_length=100)
    color: str = Field(default="#FF0000", max_length=7, pattern=r"^#[0-9A-Fa-f]{6}$")


class DefectClassOut(DefectClassCreate):
    id: int
    model_config = {"from_attributes": True}


# ---- Project ----
class ProjectCreate(BaseModel):
    name: str = Field(max_length=200)
    description: Optional[str] = None
    resize_h: int = Field(default=2048, ge=640)
    resize_w: int = Field(default=2048, ge=640)
    crop_size: int = Field(default=640, ge=320, le=8192)
    overlap: float = Field(default=0.2, ge=0.0, le=0.5)
    class_names: list[DefectClassCreate] = Field(
        default_factory=lambda: [
            DefectClassCreate(class_index=0, name="defect_1", color="#FF0000"),
            DefectClassCreate(class_index=1, name="defect_2", color="#00FF00"),
            DefectClassCreate(class_index=2, name="defect_3", color="#0000FF"),
        ]
    )


class ProjectUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    resize_h: Optional[int] = None
    resize_w: Optional[int] = None
    crop_size: Optional[int] = None
    overlap: Optional[float] = None
    status: Optional[str] = None


class ProjectOut(BaseModel):
    id: int
    name: str
    description: Optional[str]
    resize_h: int
    resize_w: int
    crop_size: int
    overlap: float
    status: str
    created_at: datetime
    updated_at: datetime
    defect_classes: list[DefectClassOut] = []
    model_config = {"from_attributes": True}


class ProjectStats(ProjectOut):
    """带统计信息的项目详情"""
    total_images: int = 0
    unlabeled_count: int = 0
    labeling_count: int = 0
    labeled_count: int = 0
    reviewed_count: int = 0
    total_annotations: int = 0

File: web/src/api/project.ts
import api from './index'

export interface DefectClass {
  id?: number
  class_index: number
  name: string
  color: string
}

export interface Project {
  id: number
  name: string
  description: string | null
  task_type: 'seg' | 'det' | 'cls' | 'obb'
  resize_h: number
  resize_w: number
  crop_size: number
  overlap: number
  status: string
  created_at: string
  updated_at: string
  defect_classes: DefectClass[]
}

export interface ProjectStats extends Project {
  total_images: number
  unlabeled_count: number
  labeling_count: number
  labeled_count: number
  reviewed_count: number
  total_annotations: number
}

export interface ProjectCreate {
  name: string
  description?: string
  task_type?: 'seg' | 'det' | 'cls' | 'obb'
  resize_h?: number
  resize_w?: number
  crop_size?: number
  overlap?: number
  class_names?: DefectClass[]
}

export interface MergeReport {
  task_type: string
  pack_images: number
  pack_annotations: number
  matched_images: number
  new_images: number
  added_annotations: number
  skipped_duplicates: number
  new_classes: string[]
  cls_conflicts: { filename: string; target_class: string; incoming_class: string }[]
  unmatched_no_image: number
  dry_run: boolean
}

type DownloadProgressHandler = (percent: number) => void

function downloadProgressConfig(onProgress?: DownloadProgressHandler) {
  return {
    responseType: 'blob' as const,
    timeout: 600000,
    onDownloadProgress: (evt: any) => {
      if (!onProgress) return
      if (evt.total) {
        onProgress(Math.min(99, Math.round((evt.loaded / evt.total) * 100)))
      } else if (evt.loaded > 0) {
        onProgress(8)
      }
    },
  }
}

export const projectApi = {
  list: () =>
    api.get<Project[]>('/projects'),

  get: (id: number) =>
    api.get<ProjectStats>(`/projects/${id}`),

  create: (data: ProjectCreate) =>
    api.post<Project>('/projects', data),

  update: (id: number, data: Partial<ProjectCreate>) =>
    api.put<Project>(`/projects/${id}`, data),

  delete: (id: number) =>
    api.delete(`/projects/${id}`),

  addClass: (projectId: number, data: DefectClass) =>
    api.post(`/projects/${projectId}/classes`, data),

  updateClass: (projectId: number, classId: number, data: DefectClass) =>
    api.put(`/projects/${projectId}/classes/${classId}`, data),

  deleteClass: (projectId: number, classId: number) =>
    api.delete(`/projects/${projectId}/classes/${classId}`),

  exportPackage: (id: number, onProgress?: DownloadProgressHandler) =>
    api.get(`/projects/${id}/export-package`, downloadProgressConfig(onProgress)),

  exportFullPackage: (id: number, onProgress?: DownloadProgressHandler) =>
    api.get(`/projects/${id}/export-full-package`, downloadProgressConfig(onProgress)),

  importPackage: (file: File) => {
    const fd = new FormData()
    fd.append('file', file)
    return api.post<{ project_id: number; project_name: string; renamed: boolean; image_count: number; annotation_count: number }>(
      '/projects/import-package',
      fd,
      { headers: { 'Content-Type': 'multipart/form-data' }, timeout: 600000 }
    )
  },

  importFullPackage: (file: File) => {
    const fd = new FormData()
    fd.append('file', file)
    return api.post<{
      project_id: number
      project_name: string
      renamed: boolean
      image_count: number
      annotation_count: number
      train_task_count: number
      epoch_log_count: number
      weight_file_count: number
    }>(
      '/projects/import-full-package',
      fd,
      { headers: { 'Content-Type': 'multipart/form-data' }, timeout: 600000 }
    )
  },

  mergePackage: (projectId: number, file: File, dryRun: boolean) => {
    const fd = new FormData()
    fd.append('file', file)
    return api.post<MergeReport>(
      `/projects/${projectId}/merge-package?dry_run=${dryRun}`,
      fd,
      { headers: { 'Content-Type': 'multipart/form-data' }, timeout: 600000 }
    )
  },
}
